### DISCUSSION TITLE
[Quest] Usage of Datastore vs SharedPreference

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#1873] [Ideas]
**Describe the issue to be researched**
- [x] Usage of Datastore vs SharedPreference
**Describe the goal of the research**
- [x] Document the Pros and Cons of the usage of Datastore over SharedPreference
**Describe the methodology**
- [x] Document the use cases for the use of either approach

**Comment by GibsonRuitiari** [comment]
If I may, both of them are used to store simple values, such as user preferences or tokens, in android. Datastore [DS] remedies most of the defects suffered by Shared Preference, this is in terms of   DS performing its functions asynchronously hence making UI Thread Safe as compared to Shared Pref. Also, DS can store multiple type of values including objects and Proto buffs whereas Shared Pref only allows primitive values. 

The one defect DS suffers from is encryption. Shared Pref uses the android sandbox by default to encrypt its values whereas DS does not do that, so values saved are always in a readable format, hence making it unsuitable for storage of sensitive data. See [Encryption issue] (https://issuetracker.google.com/issues/199770477) [as of now you may be forced to write your own encryption, AES encryption preferably, to secure your sensitive data]. Beyond usage of DS, I personally haven't tested the performance of DS, but I have with Shared Preferences, and its unfortunately, slow especially when it comes to persisting [writing] values to disk.

Side note: If you want the best of both worlds: performance, concurrency and security, then you might want to check out [MMKV](https://github.com/Tencent/MMKV) library from Tencent.

**Comment by ahsanbhatti49** [comment]
**SharedPreferences** is the common way used for storing key-value pairs in local storage. Here are some Pros of using Shared prefs.

- **Synchronous updates** — All the process run in the main thread
- **Primitive Values** — The only values available are Boolean, Int, String, or even a Set<T> where T is a primitive value
- **They can also be encrypted** — SharedPreferences are the main instrument inside the incredible library androidx.security:security-cryptothat can encrypt data using a KeyStore approach, but it also follows the same restrictions and approaches that normal SharedPreferences, so, you can only encrypt primitive values, that is more than enough for multiple tasks that requires encryption.

**Disadvantages of using SharedPreferences**

- They are not safe to call on UI thread although as they have asynchronous API that can appear to be safe to call on the UI thread, but which actually does disk I/O operations that might cause issues.
- It’s not safe from runtime exceptions as they throw parsing errors
- Cannot signal errors
- It doesn’t provide Type-safety

Lets talk  about the pros of using **Datastore** :

- **Asynchronous updates** — All the process of editing and fetching data from the schema of **DataStore** are addressed over a flow that represents the stream of values that are being asynchronously computed
- **Objects** — You can have an object inside **DataStore** with the use of **ProtoBuffers** that will make this approach type safe, without this, you will be able to store only Key-Value primitives without type safety
- **Encryption** — There’s non available encryption for the moment for **DataStore**, but we can Encrypt a full data base using **Room**.

We can refer this chart : 
![Screenshot from 2022-11-28 17-41-39](https://user-images.githubusercontent.com/18699296/204280704-f8c85f2b-08c9-4921-b455-d56d9eec4ed6.png)


References:
[https://proandroiddev.com/is-jetpack-datastore-a-replacement-for-sharedpreferences-efe92d02fcb3](url)
[https://medium.com/knowing-android/wild-storage-part-3-datastore-and-sharedpreferences-b32e8686f22b](url)
[https://proandroiddev.com/securing-androids-datastore-ad56958ca6ee](url)

**Comment by f-odhiambo** [comment]
cc @ellykits
